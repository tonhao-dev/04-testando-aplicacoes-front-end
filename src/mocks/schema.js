export default {
  fields: [
    {
      type: 'text',
      name: 'name',
      label: 'Nome',
      placeholder: 'Seu nome',
      validations: 'isRequired'
    },
    {
      type: 'email',
      name: 'email',
      label: 'Email',
      placeholder: 'seu.email.mais@top.com',
      validations: 'isEmail'
    },
  ],
  submit: 'Assinar'
};
