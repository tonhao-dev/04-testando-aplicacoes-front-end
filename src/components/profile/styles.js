import styled, { css } from 'styled-components';

export const Profile = styled.div`
  ${({ theme }) => css`
    width: 300px;
    letter-spacing: 0.8px;

    & > figure {
      width: 100%;
      height: 250px;
      display: flex;
      align-items: center;
      justify-content: center;
      color: ${theme.colors.white};
      background: ${theme.colors.post.color};

      & > * {
        width: 100%;
        height: 100%;
        object-fit: cover;
        pointer-events: none;
      }

      svg {
        max-width: 100px;
      }
    }
  `}
`;

export const Name = styled.p`
  ${({ theme }) => css`
    padding: ${theme.spacing.default};
    border-bottom: 1px solid ${theme.colors.post.color};
    text-align: center;
    font-weight: bold;
  `}
`;

export const Information = styled.div`
  ${({ theme }) => css`
    display: flex;
    flex-direction: column;
    padding: ${theme.spacing.default};
  `}
`;

export const Text = styled.p`
  ${({ theme }) => css`
    &:not(:last-child) {
      margin-bottom: ${theme.spacing.half};
    }

    & > * {
      display: block;
    }
  `}
`;

export const Title = styled.b`
  ${({ theme }) => css`
    color: ${theme.colors.terminal.blue.default};
  `}
`;

export const Data = styled.span`
  text-transform: lowercase;
`;

export const Actions = styled.div`
  ${({ theme }) => css`
    border-top: 1px solid ${theme.colors.post.color};
    padding: ${theme.spacing.default};
    display: flex;
    align-items: center;
    justify-content: space-evenly;
  `}
`;
