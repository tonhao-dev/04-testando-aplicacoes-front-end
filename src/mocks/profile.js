export const profile = {
  email: 'Lonny.Dietrich@yahoo.com',
  userName: 'Donna.Watsica38',
  name: 'Hipolito',
  lastName: 'Kassulke',
  avatar: 'https://avatars0.githubusercontent.com/u/13336905?s=460&u=2e6422672bb3f2aeaaa034e9605e953b5bbf49ca&v=4',
  role: 'USER',
  uid: '1234567-abc',
};

export const profileList = Array.from({ length: 20 }, () => profile);
