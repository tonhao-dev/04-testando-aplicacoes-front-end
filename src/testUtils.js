import { render, waitFor } from '@testing-library/react';

// Providers
import { BrowserRouter } from 'react-router-dom';
import ThemeProvider from './providers/theme';
import Providers from './providers';

// Utils
export const AnyPage = () => <div data-testid="any page" />
export const waitTimersByTime = async time => {
  await waitFor(() => {
    jest.advanceTimersByTime(time);
  });
};

// Renders
export const renderWithTheme = (ui, options) => render(ui, { wrapper: ThemeProvider, ...options });
export const renderWithRouter = (ui, options) => render(ui, { wrapper: BrowserRouter, ...options });
export const renderWithProviders = (ui, options) => render(ui, { wrapper: Providers, ...options });
