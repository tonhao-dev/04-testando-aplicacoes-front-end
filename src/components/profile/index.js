import React from 'react';
import PropTypes from 'prop-types';

import { EditOutline as EditIcon } from '@styled-icons/evaicons-outline/EditOutline';
import { Trash2Outline as DeleteIcon } from '@styled-icons/evaicons-outline/Trash2Outline';

import Avatar from '../avatar';
import Card from '../card';
import Button from '../button';

import * as S from './styles';

const Profile = ({
  email,
  userName,
  name,
  lastName,
  avatar,
  role,
  uid,
  editable,
  onClickDelete,
  onClickEdit
}) => (
  <S.Profile>
    <Card>
      <Avatar src={avatar} />
      <S.Name>
        {name} {lastName}
      </S.Name>
      <S.Information>
        <S.Text>
          <S.Title>email</S.Title>
          <S.Data>{email}</S.Data>
        </S.Text>
        <S.Text>
          <S.Title>username</S.Title>
          <S.Data>{userName}</S.Data>
        </S.Text>
        <S.Text>
          <S.Title>role</S.Title>
          <S.Data>{role}</S.Data>
        </S.Text>
      </S.Information>
      <S.Actions>
        <Button
          type="blue"
          round
          disabled={!editable}
          icon={<EditIcon />}
          onClick={() => onClickEdit(uid)}
          aria-label="editar"
        />
        <Button
          type="red"
          round
          disabled={!editable}
          icon={<DeleteIcon />}
          onClick={() => onClickDelete(uid)}
          aria-label="deletar"
        />
      </S.Actions>
    </Card>
  </S.Profile>
);

Profile.propTypes = {
  email: PropTypes.string.isRequired,
  userName: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  lastName: PropTypes.string.isRequired,
  avatar: PropTypes.string,
  role: PropTypes.string.isRequired,
  editable: PropTypes.bool,
  uid: PropTypes.string.isRequired,
  onClickDelete: PropTypes.func,
  onClickEdit: PropTypes.func,
};

Profile.defaultProps = {
  avatar: null,
  editable: false,
  onClickDelete: () => {},
  onClickEdit: () => {},
};

export default Profile;
