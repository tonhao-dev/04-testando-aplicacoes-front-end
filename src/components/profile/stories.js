import React from 'react';

import Profile from '.';
import { profile } from '../../mocks/profile';

export default {
  title: 'Components/Profile',
  component: Profile,
  args: {
    ...profile
  }
};

const Template = (args) => <Profile {...args} />;

export const Default = Template.bind({});
Default.args = {
  editable: false
};

export const Editable = Template.bind({});
Editable.args = {
  editable: true
};
