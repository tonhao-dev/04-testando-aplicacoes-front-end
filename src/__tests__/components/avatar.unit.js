// Utilitários
import { screen } from '@testing-library/react';
import { renderWithTheme } from '../../testUtils';

// Componente
import Avatar from '../../components/avatar';

// Mocks
import { profile } from '../../mocks/profile';

describe('<Avatar />', () => {
  it('Deve renderizar um ícone ao não possuir uma imagem', () => {
    renderWithTheme(<Avatar />);

    const figura = screen.getByRole('figure');
    const imagem = screen.queryByRole('img');

    expect(figura).toBeInTheDocument();
    expect(imagem).not.toBeInTheDocument();
    expect(figura).toMatchSnapshot();
  });

  it('Deve renderizar uma imagem quando disponível', () => {
    renderWithTheme(<Avatar src={profile.avatar} />);

    const figura = screen.getByRole('figure');
    const imagem = screen.queryByRole('img');

    expect(figura).toBeInTheDocument();
    expect(imagem).toBeInTheDocument();
    expect(figura).toMatchSnapshot();
  });
});
