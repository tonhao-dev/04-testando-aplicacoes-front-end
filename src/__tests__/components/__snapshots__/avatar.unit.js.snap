// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Avatar /> Deve renderizar um ícone ao não possuir uma imagem 1`] = `
<figure
  class="sc-hKgILt fZnRJR"
>
  <svg
    aria-hidden="true"
    class="StyledIconBase-ea9ulj-0 cKCFNq"
    fill="currentColor"
    focusable="false"
    viewBox="0 0 8 8"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M4.47 0c-.85 0-1.48.26-1.88.66-.4.4-.54.9-.59 1.28l1 .13c.04-.25.12-.5.31-.69C3.5 1.19 3.8 1 4.47 1c.66 0 1.02.16 1.22.34.2.18.28.4.28.66 0 .83-.34 1.06-.84 1.5-.5.44-1.16 1.08-1.16 2.25V6h1v-.25c0-.83.31-1.06.81-1.5.5-.44 1.19-1.08 1.19-2.25 0-.48-.17-1.02-.59-1.41C5.95.2 5.31 0 4.47 0zm-.5 7v1h1V7h-1z"
    />
  </svg>
</figure>
`;

exports[`<Avatar /> Deve renderizar uma imagem quando disponível 1`] = `
<figure
  class="sc-hKgILt fZnRJR"
>
  <img
    alt="avatar"
    src="https://avatars0.githubusercontent.com/u/13336905?s=460&u=2e6422672bb3f2aeaaa034e9605e953b5bbf49ca&v=4"
  />
</figure>
`;
